gpu-aware-scheduling/pkg/gpuscheduler/scheduler.go

```diff
@@ -0,0 +1,604 @@
+
+	// select GPUs. Trivial implementation selects first suitable GPUs
+	containerRequests := containerRequests(pod)
+	containerDelimeter := ""
+
+	for i, containerRequest := range containerRequests {
+		cards, err := m.getCardsForContainerGPURequest(containerRequest,
+			perGPUCapacity, nodeName, podName, nodeResourcesUsed, gpuMap)
+		if err != nil {
+			klog.Errorf("container %v out of %v did not fit", i+1, len(containerRequests))
+
+			return "", err
+		}
+
+		annotation += containerDelimeter
+		cardDelimeter := ""
+
+		for _, card := range cards {
+			annotation += cardDelimeter + card
+			cardDelimeter = ","
+		}
+
+		containerDelimeter = "|"
+	}
+
+	return annotation, nil
+}
+
+// checkResourceCapacity returns true if the needed resources fit based on capacity and used resources.
+func checkResourceCapacity(neededResources, capacity, used resourceMap) bool {
+	for resName, resNeed := range neededResources {
+		if resNeed < 0 {
+			klog.Error("negative resource request")
+
+			return false
+		}
+
+		resCapacity, ok := capacity[resName]
+		if !ok || resCapacity <= 0 {
+			klog.V(L4).Info(" no capacity available for ", resName)
+
+			return false
+		}
+
+		resUsed := used[resName] // missing = 0, default value is ok
+
+		if resUsed < 0 {
+			klog.Error("negative amount of resources in use")
+
+			return false
+		}
+
+		klog.V(L4).Info(" resource ", resName, " capacity:", strconv.FormatInt(resCapacity, 10), " used:",
+			strconv.FormatInt(resUsed, 10), " need:", strconv.FormatInt(resNeed, 10))
+
+		if resUsed+resNeed < 0 {
+			klog.Error("resource request overflow error")
+
+			return false
+		}
+
+		if resCapacity < resUsed+resNeed {
+			klog.V(L4).Info(" not enough resources")
+
+			return false
+		}
+	}
+
+	klog.V(L4).Info(" there is enough resources")
+
+	return true
+}
+
+func (m *GASExtender) bindNode(args *extender.BindingArgs) *extender.BindingResult {
+	result := extender.BindingResult{}
+
+	// first get the pod
+	pod, err := m.cache.fetchPod(args.PodNamespace, args.PodName)
+	if err != nil {
+		klog.Warningf("Pod %s couldn't be read or pod vanished", args.PodName)
+
+		result.Error = err.Error()
+
+		return &result
+	}
+
+	m.rwmutex.Lock()
+	klog.V(L5).Infof("bind %v:%v to node %v locked", args.PodNamespace, args.PodName, args.Node)
+	defer m.rwmutex.Unlock()
+
+	// pod should always fit, but one never knows if something bad happens between filtering and binding
+	annotation, err := m.getAnnotationForPodGPURequest(pod, args.Node)
+	resourcesAdjusted := false
+
+	if err == nil {
+		klog.V(L3).Infof("bind %v:%v to node %v annotation %v", args.PodNamespace, args.PodName, args.Node, annotation)
+		// adjust (add) resources in cache, this includes sanity check for the inputs
+		err = m.cache.adjustPodResourcesL(pod, add, annotation, args.Node)
+
+		if err == nil {
+			resourcesAdjusted = true
+			// annotate POD with per-container GPU selection
+			err = m.annotatePodBind(annotation, pod)
+
+			if err == nil {
```

Comment:
Can this be a != nil instead? I find it harder to follow this code with so many indentation levels.

Reply:
Point taken, but with this logic it can't be a !=.

The idea there is that at the event of first error, a graceful exit with resources returned to the book keeping is executed, and there is only one such place for cleanup. If I reverse that then there needs to be multiple cleanup chunks doing the same thing.

Reply:
consider deferring a handler